plugins/Utils/windowstatestorage.cpp

```diff
@@ -16,167 +17,319 @@
+        return QRect();
+    }
 
-    saveValue(queryString);
 }
 
-void WindowStateStorage::saveStage(const QString &appId, int stage)
+void AsyncQuery::saveStage(const QString &appId, int stage)
 {
-    const QString queryString = QStringLiteral("INSERT OR REPLACE INTO stage (appId, stage) values ('%1', '%2');")
-            .arg(sanitiseString(appId))
-            .arg(stage);
+    QSqlDatabase connection = QSqlDatabase::database(m_connectionName);
+    QSqlQuery query(connection);
+    query.prepare(m_saveStageQuery);
+    query.bindValue(":appId", appId);
+    query.bindValue(":stage", stage);
+    if (!query.exec()) {
+        logSqlError(query);
+    }
+}
 
-    saveValue(queryString);
+int AsyncQuery::getStage(const QString &appId) const
+{
+    QSqlDatabase connection = QSqlDatabase::database(m_connectionName);
+    QSqlQuery query(connection);
+    query.prepare(m_getStageQuery);
+    query.bindValue(":appId", appId);
+    query.exec();
+    if (!query.isActive() || !query.isSelect()) {
+        logSqlError(query);
+        return -1;
+    }
+    if (!query.first()) {
+        return -1;
+    }
+    bool converted = false;
+    QVariant resultStr = query.value(0);
+    int result = resultStr.toInt(&converted);
+    if (converted) {
+        return result;
+    } else {
+        qWarning() << "getStage result expected integer, got " << resultStr;
+        return -1;
+    }
 }
 
-int WindowStateStorage::getStage(const QString &appId, int defaultValue) const
+void AsyncQuery::logSqlError(const QSqlQuery query) const
 {
-    const QString queryString = QStringLiteral("SELECT stage FROM stage WHERE appId = '%1';")
-            .arg(sanitiseString(appId));
+    qWarning() << "Error executing query" << query.lastQuery()
+               << "Driver error:" << query.lastError().driverText()
+               << "Database error:" << query.lastError().databaseText();
+}
 
-    QSqlQuery query = getValue(queryString);
+const QString AsyncQuery::getDbName()
+{
+    QSqlDatabase connection = QSqlDatabase::database(m_connectionName);
+    return connection.databaseName();
+}
 
-    if (!query.first()) {
-        return defaultValue;
+WindowStateStorage::WindowStateStorage(const QString& dbName, QObject *parent):
+    QObject(parent),
+    m_thread()
+{
+    QString dbFile;
+    if (dbName == nullptr) {
+        const QString dbPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + QStringLiteral("/unity8/");
+        QDir dir;
+        dir.mkpath(dbPath);
+        dbFile = QString(dbPath) + QStringLiteral("windowstatestorage.sqlite");
+    } else {
+        dbFile = dbName;
+    }
+    m_asyncQuery = new AsyncQuery(dbFile);
+    m_asyncQuery->moveToThread(&m_thread);
+    connect(&m_thread, &QThread::finished, m_asyncQuery, &QObject::deleteLater);
+    m_thread.start();
+    bool queryInitSuccessful;
+    QMetaObject::invokeMethod(m_asyncQuery, "initdb",
+                              Qt::BlockingQueuedConnection,
```

Comment:
Does this need to be blocking? And if it some of the funcs get called before its inited, maybe use QWaitCondition or something in the funcs to make sure its inited,

Reply:
Well, I suppose it would make Lomiri's start a little faster under conditions of high IO load... but the first application that the user launched would still have to wait on the database to initialize. In other words, you're just moving the delay to first app launch instead of Lomiri launch. Is that worth it?

Reply:
Hmm, it can be some time between this class is init and the user starts an app, so it could be an overall win. I'm not sure about this though, and the complexity factor might make this idea not worth it.

Reply:
A Queued connection is enough, really:

> It needs to be synchronous, else it is possible to save a value before the connection is opened. We don't really want that.

This cannot happen, because the thread is running an event loop, and will process all the queued requests in the same order they were sent: so it will always process the `initdb()` method first.